### ISSUE TITLE
Prepare Istio 1.16

### ISSUE THREAD

**Issue by dhawton** [istio/istio#41015]
Modified version of #40127 (from 1.15)

---

This issue aims to track all steps needed to prepare the release 1.16 branches/builds.

This issue was created early to help document the steps using some new automation as well as any ordering changes, etc. Please update/add additional steps taken here so we can document for future releases.

See https://github.com/istio/istio/issues/24473 for steps used for the release 1.7 branch cutting; (note: 1.8 will be more relevant, see https://github.com/istio/istio/issues/27993). Note that was before automation, but does have pointers to PRs to see what some of the steps did. **Note that the actual coding steps done by the automation steps are left in this issue for reference only. You only need to run the automated steps, and not the individual instructions. Also note that the reference steps may be INCOMPLETE as the automation is updated, not necessarily the reference steps.**

For now, the code to run is in https://github.com/istio/release-builder master branch. From within that branch, one can do a `make shell` and then within the shell run `REPO_ORG=myorg STEP=x ./release/branch.sh` (you can update the STEP in the `trigger-branch` file and not set it manually). This generates the changes for review (try`git status` or `git diff`).  Adding DRY_RUN=false will cause the PRs to be created in the REPO_ORG.

Note that the automated steps 3 and on require the new release's branch in the various repos which are created by automated step 2.

- [x] Set up release manager team - Make sure the new team has write access to the repositories. [PR](https://github.com/istio/community/pull/793)
- [x] Send an announcement on Discuss pre-announcing branch cut (adding, but are branch cut notifications needed and where?)
- [x] Send an announcement on Discuss announcing the start of branch cut and to stop merging of PRs (adding, but are branch cut notifications needed and where?)

- [x] (**Automation step=1**) Update dependencies. This must be done **before branching** so that PRs can be easily backported later. [PR](https://github.com/istio/istio/pull/41466)
  - [x] run `go get github.com/envoyproxy/go-control-plane@main` to update
  - [x] Run the following command in istio/istio `UPDATE_BRANCH=master ./bin/update_deps.sh; make gen` ([example](https://github.com/istio/istio/pull/36812))
  - [x] Merge PR in istio/istio. Wait for it to complete. 

- [x]  (**Automation step=2**)  (new step because we are waiting on prior change to master before branching) Create a `release-1.16` branch in every Istio repo - run commands `export org=xyz ; export repo=xyz ; (git clone git@github.com:${org}/${repo}.git && cd $repo && git checkout -b release-1.16 && git push --set-upstream origin release-1.16)`.
  - [x] istio/istio 
  - [x] istio/pkg
  - [x] istio/api 
  - [x] istio/envoy  (**NOT done as part of release-builder**) ([ ] Should update code to include this again?) 
  - [x] istio/proxy
  - [x] istio/client-go 
  - [x] istio/tools
  - [x] istio/common-files 
  - [ ] istio/gogo-genproto (**No longer used and not done as part of release-builder [1.15 PR](https://github.com/istio/release-builder/pull/1003)**) 
  - [x] istio/release-builder 
  - [x] istio/enhancements
  - Explicitly skipped: istio-releases/pipeline, istio/tests, istio/istio.io, istio/test-infra
  - [x] No PRs to merge as the new branches were created by automation

- [ ] Set up branch protection. Modify prow/config.yaml in test-infra to add release manages as owners for a branch. [1.16 PR](https://github.com/istio/test-infra/pull/4336)

- [x] (**Automation/Not very automated step=3**) (doesn't work locally, appears to need credential to add a tag to an existing build-tools image)(new step because this might need to be done separately) Set up prow on release branches `go run generate.go branch 1.16` (Note: Needed to run postsubmit in next step) (Contact Googler or ping in #test-and-release slack channel for help). [1.16 PR](https://github.com/istio/test-infra/pull/4339)
  - Manual steps if Googler can't run automation:
    - [x] Have Googler or someone with creds tag new build-tools, build-tools-proxy and build-tools-centos images, for example (replace hash with latest):
      - [x] `gcloud container images add-tag gcr.io/istio-testing/build-tools:master-6363b97c6bc60b7a3a8d52c5e08322d8394149e2 gcr.io/istio-testing/build-tools:release-1.16-6363b97c6bc60b7a3a8d52c5e08322d8394149e2`
      - [x] `gcloud container images add-tag gcr.io/istio-testing/build-tools-proxy:master-6363b97c6bc60b7a3a8d52c5e08322d8394149e2 gcr.io/istio-testing/build-tools-proxy:release-1.16-6363b97c6bc60b7a3a8d52c5e08322d8394149e2`
      - [x] `gcloud container images add-tag gcr.io/istio-testing/build-tools-centos:master-6363b97c6bc60b7a3a8d52c5e08322d8394149e2 gcr.io/istio-testing/build-tools-centos:release-1.16-6363b97c6bc60b7a3a8d52c5e08322d8394149e2`
    - [x] Create new branch off of latest test-infra master branch
    - [x] Run `make shell`
    - [x] You'll need to run a couple go get's for the tooling: `go get github.com/kballard/go-shellquote` `go get istio.io/test-infra/tools/prowgen/pkg` -- don't commit the modified go.mod and go.sum from these commands
    - [x] Generate the jobs: `go run tools/prowgen/cmd/prowgen/main.go branch 1.16`
    - [x] Generate the private jobs: `go run ./tools/generate-transform-jobs/main.go branch 1.16`
    - [x] In the job configs prow/**/*-1.16.yaml, replace any image build-tools(-centos|-proxy):master with build-tools(-centos|-proxy):release-1.16
    - [x] Verify there is no duplicated env entries of the following in the configs:
      ```
      - name: BUILD_WITH_CONTAINER
        value: "0"
      ```
    - [x] Manually update `testgrid/config.yaml` manually, see last PR for entries to add and substitute the release (there is tooling to do this, just not sure where yet)
    - [x] Run `make gen`

- [x] (**Automation step=4**) (new step since we want the automation from prior step to actually create the image after PR merges) PRS: [1 - istio/pkg](https://github.com/istio/pkg/pull/651) [2 - istio/client-go](https://github.com/istio/client-go/pull/1175) [3 - istio/tools](https://github.com/istio/tools/pull/2179) [4 - istio/common-files](https://github.com/istio/common-files/pull/688) [5 - istio/enhancements](https://github.com/istio/enhancements/pull/126) [6 - istio/release-builder](https://github.com/istio/release-builder/pull/1238) [7 - istio/istio](https://github.com/istio/istio/pull/41521) [8 - istio/api](https://github.com/istio/api/pull/2512) [9 - istio/proxy](https://github.com/istio/proxy/pull/4124) [10 - istio/envoy](https://github.com/istio/envoy/pull/337)
  - [x] Updates istio/tools to build new release-1.16 build image (update BRANCH in docker/build-tools/build-and-push.sh. PR postsubmit will create new container images whose name will be used in the next step) 
  - Update common-files in _new release_. You first have to manually update the _common/Makefile.common.mk_ *update-common* target to point to the _new release_  (but not in common-files). The next step will cause the prow automation to actually call `make update-common` in these repos to do the actual `make update-common` against the new common-files release branch
    - [ ] istio/common-files - skipped. Covered in next step
    - [x] istio/istio 
    - [x] istio/pkg 
    - [x] istio/api
    - [x] istio/client-go
    - [ ] istio/gogo-genproto. (**No longer used**)
    - [x] istio/proxy 
    - [x] istio/release-builder 
    - [x] istio/tools 
  - Update `CODEOWNERS` to contain only the release managers for this release. Command: `export org=xyz ; export repo=xyz ; (git clone git@github.com:${org}/${repo}.git && cd $repo && git checkout release-1.16 && git checkout -b release-1.16-codeowners && echo '* @istio/release-managers-1.16' > CODEOWNERS && git add CODEOWNERS && git commit -m 'Set release managers as CODEOWNERS for release-1.16' && git push --set-upstream origin release-1.16-codeowners)`.
    - [x] istio/istio 
    - [x] istio/pkg 
    - [x] istio/api 
    - [x] istio/proxy
    - [x] istio/client-go 
    - [x] istio/tools
    - [x] istio/common-files
    - [ ] istio/gogo-genproto (**No longer used**)
    - [x] istio/release-builder
    - [x] istio/enhancements
  - [x] Stop publishing `latest` tags
  - Update istio/release-builder branch changes from `master` to new release in build.sh, publish.sh and manifests. 
- [x] Merge PRs from STEP 4. Wait for the new build images to be created at https://gcr.io/istio-testing/build-tools.

- [x] **WARNING -- DO NOT MERGE** As part of the STEP 4 PR post-submits, a new PR will be created in the istio/common-files repo with a title like: `Automator: update build-tools image@release-1.16 in istio/common-files@release-1.16`. The work in this PR is only a portion of Step 5, and merging it will cause issues with reverting the repos to the main branch common-files. **CLOSE this PR**.

- [x] (**Automation step=5**) (new step since we need image from prior step) Update istio/common-files to set release-1.16 build image (Update the UPDATE_BRANCH in files/common/Makefile.common.mk to be the new release name, and IMAGE_VERSION in files/common/scripts/setup_env.sh  to be the new build image from prior step (found at https://gcr.io/istio-testing/build-tools). [1.16 PR](https://github.com/istio/common-files/pull/691)

- [x] Merge PR from STEP 5. Wait until step 5 PR merges and all the postsubmit generated PRs (to update common files in repos) merge.

Additional steps that still need automation (if possible)
- [x] Add new label: cherrypick/release-1.16:
- [x] Create a 1.16 milestone:
- [x] Fix up proxy automated job to pull from correct envoyproxy release branch if possible (set UPDATE_BRANCH). [example](https://github.com/istio/test-infra/pull/3795) (** NOTE there may need to be a revert of a PR for updates in main branch post Envoy 1.23.**) [1.16 PR](https://github.com/istio/test-infra/pull/4345)
- [x] Also, a manual update can be done to fix up the proxy and istio repos to point to latest in that branch. [example](https://github.com/istio/proxy/pull/3663)

- [ ] Bump master to version 1.16 in the release-1.16 branches
  - [ ] Verify UPDATE_BRANCH is release-1.16 and not master
  - [ ] Others like base image, etc.

- [x] Bump 1.16 to 1.17 in the master branch 
  - [x] Update Istio: 1.17.0 to 1.17.0, 1.17-dev to 1.17-dev, etc: [1.16 PR](https://github.com/istio/istio/pull/41552)

- [x] Send an announcement on Discuss announcing branch cut complete and PRs can be merged again (adding, but are branch cut notifications needed and where?)

- [x] update deps before building beta

- [ ] Publish a beta release (from https://docs.google.com/document/d/1ilwz-HGSeiRDCFwP5rm-5d_UY6OOYL_nflx7bNggek0/edit#heading=h.qex63c29z2to, builds on release branch are labeled `beta`. Updating below names to reflect that.
  - [x] Run bin/update_deps.sh in istio/istio
  - [x] Trigger a 1.16 beta.0 build
  - [x] Some verification to verify build is good
  - [ ] Publish a 1.16 beta.0 release
- [ ] Send an announcement on Discuss announcing availability of beta. May wait and add something about starting the first round of community testing. (Done in Slack)
- [x] Ask istio.io team to run job to update to use the 1.16 branch (probably @ericvn) [example](https://github.com/istio/istio.io/pull/10455)
- [ ] Run the long-running tests on the dev builds

**Comment by shivanshu1333** [comment]
Interested in picking up some work...
